### PR TITLE
Scale tests with cluster size

### LINKED ISSUE #483: Configure number of workers and instance types in A/B tests
At the moment, all of our tests use hardcoded number of VMs and instance types - 10 x m6i.large for all tests that use the `small_client` fixture.

This leaves us blind to three major use cases:
- workers with more than 2 threads
- clusters with 20+ workers - this triggers completely different code paths in the scheduler
- clusters with many workers (100~10,000) - this can highlight problems with big-O complexity in the scheduler algorithms.

# Proposed design
It would be extremely costly to enable the above in the nightly tests, so I propose to add a switch to A/B tests and trigger them as needed.

Require new config files `AB_environments/AB_<name>.cluster.yaml`, with the following config:
```yaml
# These are arbitrary kwargs to coiled.Cluster

# For all tests using the small_client fixture
small_cluster:
  n_workers: 10
  worker_vm_types: [m6i.large]  # 2CPU, 8GiB
  scheduler_vm_types: [m6i.xlarge]

# For test_parquet.py
parquet_cluster:
  n_workers: 15
  worker_vm_types: [m5.xlarge]  # 4 CPU, 16 GiB
  scheduler_vm_types: [m5.xlarge]
```

and have them reach the respective fixtures.
Note that the above will not impact special tests which create their own clusters, e.g. test_spill.py.

Review all tests to verify that the test workload scales with the total amount of CPUs or memory in the cluster, so that you get a meaningful workload on a 10 workers cluster as well as on a 200 workers one.

Review which tests hit the spill threshold and, 
  - if spilling isn't an integral part of the test, make sure the test data is sized so that spilling is consistently avoided across various cluster sizes and threads per worker
  - (harder) if spilling IS an integral part of the test (read: shuffle), make sure the test data is sized so that spilling is consistently hit in roughly the same proportion

Note that, even with the review in dataset scaling above, it is expected that testing e.g. a 20-threads cluster vs. a 40-threads cluster is going to be comparing apples and oranges - namely for all tests that implement an algorithm that doesn't scale linearly with the number of chunks. However it will be very interesting to compare eg. how  a 10 workers, 2 threads-per-worker cluster compares to a 5 workers, 4 threads-per-worker one.

### PR DESCRIPTION
**Pull request by crusaderky**
Partially closes #483 

Review all tests so that they retain a pseudo-constant runtime as the cluster size goes up. Disable those that can't for very large clusters.

Demo with 2x[1] and 5x[2] cluster sizes:
https://github.com/coiled/coiled-runtime/actions/runs/3584320320
[1] 20 workers on small_cluster, 30 workers on parquet_cluster
[2] 50 workers on small_cluster, 75 workers on parquet_cluster

### CODE REVIEW

**Comment by ntabris** [comment]
Often I want to run a consistent workload on different sized clusters, in order to compare cost. Please make sure that's still easy, and that it's not too hard to tell for given test run what the exact workload was. (FWIW I'd lean toward parameterizing tests rather than have test just internally decide to do different things on different hardware.)

**Comment by crusaderky** [comment]
> Please make sure that's still easy

**Most** tests in main are already scaling their workload to the cluster size.
This PR just makes it organic throughout **all** tests.

File: tests/benchmarks/test_work_stealing.py [hunk 1]
```diff
@@ -10,14 +10,18 @@
...
+@run_up_to_nthreads("small_cluster", 50, reason="fixed dataset")
 def test_trivial_workload_should_not_cause_work_stealing(small_client):
     root = delayed(lambda n: "x" * n)(utils.parse_bytes("1MiB"), dask_key_name="root")
     results = [delayed(lambda *args: None)(root, i) for i in range(10000)]
     futs = small_client.compute(results)
     small_client.gather(futs)
 
 
+@run_up_to_nthreads("small_cluster", 50, reason="fixed dataset")
 @pytest.mark.xfail(
```
Comment (ncclementi) [inline tests/benchmarks/test_work_stealing.py:25]: Since we are working with versions  `> 2022.11.1`  should we remove this. We can do it in a separate PR if it's out of scope.

**Review by ncclementi — APPROVED** [review]
This LGTM. 
I have one comment/question for learning purposes: How was the choice made for the max value the n_threads (50 or 100) in each case?

File: tests/benchmarks/test_work_stealing.py (hunk 1 above)
Comment (crusaderky) [reply tests/benchmarks/test_work_stealing.py:25]: Yes definitely out of scope

**Comment by crusaderky** [comment]
> I have one comment/question for learning purposes: How was the choice made for the max value the n_threads (50 or 100) in each case?

It was from observing that there was virtually no change between 40 and 100 threads
